### PR TITLE
add sql level test for decimal data type

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

extract and add test for decimal data type in the SQL level.

related #122

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by jackwener — APPROVED** [review]
Great job❤

It cover
- lt/lteq/...  six compare operator
- min/max
- group
- sort

File: datafusion/core/tests/decimal_data.csv [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+0.00003,0.000000000003,5,true,0.000011
+0.00004,0.000000000004,5,true,0.000044
+0.00004,0.000000000004,12,false,0.000040
+0.00004,0.000000000004,14,true,0.000040
+0.00004,0.000000000004,8,false,0.000044
+0.00005,0.000000000005,9,true,0.000052
+0.00005,0.000000000005,4,true,0.000078
+0.00005,0.000000000005,8,false,0.000033
+0.00005,0.000000000005,100,true,0.000068
+0.00005,0.000000000005,1,false,0.000100
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/tests/decimal_data.csv:16]: https://github.com/apache/arrow-datafusion/pull/2224

File: datafusion/core/tests/sql/decimal.rs [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    Ok(())
+}
+
+#[tokio::test]
+async fn decimal_agg_function() -> Result<()> {
+    let ctx = SessionContext::new();
+    register_decimal_csv_table_by_sql(&ctx).await;
+    // min
+    let sql = "select min(c1) from decimal_simple where c4=false";
+    let actual = execute_to_batches(&ctx, sql).await;
```
Comment (alamb) [inline datafusion/core/tests/sql/decimal.rs:146]: For anyone else following along, `decimal_group_function` covers grouping by decimal columns 👍 

Something like
```sql
select max(c1) from decimal_simple group by c2;
```

File: datafusion/core/tests/sql/decimal.rs [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+
+#[tokio::test]
+async fn decimal_logic_op() -> Result<()> {
+    let ctx = SessionContext::new();
+    register_decimal_csv_table_by_sql(&ctx).await;
+    // logic operation: eq
+    let sql = "select * from decimal_simple where c1=0.00002";
+    let actual = execute_to_batches(&ctx, sql).await;
+    assert_eq!(
+        &DataType::Decimal(10, 6),
```
Comment (alamb) [inline datafusion/core/tests/sql/decimal.rs:219]: is there a reason this test only checks the schema of `c1` and not `c2`? 

It also isn't clear to me why this test is checkout the output schema but `decimal_by_filter` isn't (not that this is a problem I am just curious)

File: datafusion/core/tests/sql/decimal.rs [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+        "| 0.000050 | 0.000000000005 | 9   | true  | 0.0000520 |",
+        "| 0.000050 | 0.000000000005 | 4   | true  | 0.0000780 |",
+        "| 0.000050 | 0.000000000005 | 8   | false | 0.0000330 |",
+        "| 0.000050 | 0.000000000005 | 100 | true  | 0.0000680 |",
+        "| 0.000050 | 0.000000000005 | 1   | false | 0.0001000 |",
+        "+----------+----------------+-----+-------+-----------+",
+    ];
+    assert_batches_eq!(expected, &actual);
+
+    // logic operation: gteq
```
Comment (alamb) [inline datafusion/core/tests/sql/decimal.rs:304]: Is it worth checking somewhere what happens with different decimal scales?

Something like
```sql 
select * from decimal_simple where c1 >= CAST(0.00002 as Decimal(10,8))
```

I am not sure if the coercion to an appropriate decimal type is covered elsewhere

File: datafusion/core/tests/sql/decimal.rs [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+        "| 0.000050 | 0.000000000005 | 1   | false | 0.0001000 |",
+        "| 0.000040 | 0.000000000004 | 5   | true  | 0.0000440 |",
+        "| 0.000040 | 0.000000000004 | 12  | false | 0.0000400 |",
+        "| 0.000040 | 0.000000000004 | 14  | true  | 0.0000400 |",
+        "| 0.000040 | 0.000000000004 | 8   | false | 0.0000440 |",
+        "+----------+----------------+-----+-------+-----------+",
+    ];
+    assert_batches_eq!(expected, &actual);
+
+    let sql = "select * from decimal_simple where c1 < 0.00003 order by c1 desc,c4";
```
Comment (alamb) [inline datafusion/core/tests/sql/decimal.rs:374]: 👍

**Review by alamb — APPROVED** [review]
😍  -- looks great to me. Thank you @liukun4515

**Review by yjshen — APPROVED** [review]

File: datafusion/core/tests/sql/decimal.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:146]: I just want to cover the agg functions which support the decimal data type.

Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:146]: if other agg functions support the decimal data type, we can add the test in the function or scope.

File: datafusion/core/tests/sql/decimal.rs (hunk 3 above)
Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:219]: thanks for your point.
I will add the checker for the output schema in all test cases.

File: datafusion/core/tests/sql/decimal.rs (hunk 4 above)
Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:304]: good point.
c1 is decimal(10,6), and the cast(xxx as decimal(10,8)) is decimal(10,8), the result type should be the `decimal(12,8)`.
I will add this test case.

Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:304]: I find the bug in the precision and scale for logic comparison operation.
```
❯ explain select * from food where a >= CAST(0.00002 as Decimal(10,8));
+---------------+----------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                     |
+---------------+----------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: #food.a, #food.b, #food.c                                                                    |
|               |   Filter: #food.a >= Decimal128(Some(2000),10,8)                                                         |
|               |     TableScan: food projection=Some([0, 1, 2]), partial_filters=[#food.a >= Decimal128(Some(2000),10,8)] |
| physical_plan | ProjectionExec: expr=[a@0 as a, b@1 as b, c@2 as c]                                                      |
|               |   CoalesceBatchesExec: target_batch_size=4096                                                            |
|               |     FilterExec: CAST(a@0 AS Decimal(10, 8)) >= Some(2000),10,8                                           |
|               |       RepartitionExec: partitioning=RoundRobinBatch(16)                                                  |
|               |         CsvExec: files=[aggregate_simple.csv], has_header=false, limit=None, projection=[a, b, c]        |
|               |                                                                                                          |
+---------------+----------------------------------------------------------------------------------------------------------+
```
@alamb

File: datafusion/core/tests/sql/decimal.rs (hunk 3 above)
Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:219]: > is there a reason this test only checks the schema of c1 and not c2?
this test just wants to verify the decimal data type, I will add the test for c5.

File: datafusion/core/tests/sql/decimal.rs (hunk 4 above)
Comment (liukun4515) [reply datafusion/core/tests/sql/decimal.rs:334]: https://github.com/apache/arrow-datafusion/issues/2232

**Review by Dandandan — APPROVED** [review]

File: datafusion/core/tests/sql/decimal.rs (hunk 4 above)
Comment (alamb) [reply datafusion/core/tests/sql/decimal.rs:334]: > I find the bug in the precision and scale for logic comparison operation.

Testing for the win!

**Pull request merged**